File: dashboard-files/src/scripts/removeDuplicates.js
const fs = require('fs');
const path = require('path');

// Read the mockData.ts file
const mockDataPath = path.join(process.cwd(), 'data', 'mockData.ts');
const content = fs.readFileSync(mockDataPath, 'utf8');

// Extract the articles array
const articlesStart = content.indexOf('export const mockArticles: Article[] = [');
const articlesEnd = content.lastIndexOf('];', content.indexOf('export const publishers'));

if (articlesStart === -1) {
  console.error('Could not find mockArticles array');
  process.exit(1);
}

// Get everything before and after the articles array
const beforeArticles = content.substring(0, articlesStart);
const afterArticlesIndex = content.indexOf('];', articlesStart) + 2;
const afterArticles = content.substring(afterArticlesIndex);

// Extract just the articles JSON content
const articlesContent = content.substring(articlesStart + 'export const mockArticles: Article[] = '.length, afterArticlesIndex - 1);

// Parse the articles
let articles;
try {
  articles = eval(articlesContent); // Using eval since it's a TS file
} catch (e) {
  console.error('Error parsing articles:', e.message);
  process.exit(1);
}

console.log(`Total articles before deduplication: ${articles.length}`);

// Track duplicates
const seen = new Map();
const duplicates = [];
const unique = [];

articles.forEach((article, index) => {
  if (seen.has(article.id)) {
    duplicates.push({ id: article.id, title: article.title, index });
  } else {
    seen.set(article.id, true);
    unique.push(article);
  }
});

console.log(`\nFound ${duplicates.length} duplicate articles:`);
duplicates.forEach(dup => {
  console.log(`  - ID ${dup.id}: "${dup.title}" (index ${dup.index})`);
});

console.log(`\nUnique articles: ${unique.length}`);

// Write the deduplicated data back
const newContent = beforeArticles + 
  'export const mockArticles: Article[] = ' + 
  JSON.stringify(unique, null, 2) + 
  ';\n\n' + 
  afterArticles;

fs.writeFileSync(mockDataPath, newContent, 'utf8');
console.log('\nDuplicates removed successfully!');

File: dashboard-files/src/scripts/removeDuplicates.py
#!/usr/bin/env python3
import json
import re

# Read the mockData.ts file
with open('data/mockData.ts', 'r', encoding='utf-8') as f:
    content = f.read()

# Find the start and end of the articles array
articles_start = content.find('export const mockArticles: Article[] = [')
if articles_start == -1:
    print("Could not find mockArticles array")
    exit(1)

# Find the end of the array (looking for the closing ];)
array_content_start = articles_start + len('export const mockArticles: Article[] = ')
array_end = content.find('];', array_content_start)
if array_end == -1:
    print("Could not find end of mockArticles array")
    exit(1)

# Extract parts
before_articles = content[:array_content_start]
articles_json_str = content[array_content_start:array_end + 1]
after_articles = content[array_end + 2:]

# Parse the JSON array
try:
    articles = json.loads(articles_json_str)
except json.JSONDecodeError as e:
    print(f"Error parsing JSON: {e}")
    exit(1)

print(f"Total articles before deduplication: {len(articles)}")

# Track duplicates
seen_ids = {}
duplicates = []
unique_articles = []

for idx, article in enumerate(articles):
    article_id = article.get('id')
    if article_id in seen_ids:
        duplicates.append({
            'id': article_id,
            'title': article.get('title', 'No title'),
            'index': idx,
            'first_seen': seen_ids[article_id]
        })
        print(f"  Duplicate found: ID {article_id} at index {idx} (first seen at index {seen_ids[article_id]})")
    else:
        seen_ids[article_id] = idx
        unique_articles.append(article)

print(f"\nFound {len(duplicates)} duplicate articles")
print(f"Unique articles: {len(unique_articles)}")

# Write back the deduplicated data
new_articles_json = json.dumps(unique_articles, indent=2, ensure_ascii=False)
new_content = before_articles + new_articles_json + ';\n\n' + after_articles

with open('data/mockData.ts', 'w', encoding='utf-8') as f:
    f.write(new_content)

print("\n✓ Duplicates removed successfully!")
print(f"  Before: {len(articles)} articles")
print(f"  After: {len(unique_articles)} articles")
print(f"  Removed: {len(duplicates)} duplicates")
